### PR TITLE
Split reconcile into prepare and reconcile

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

The current "reconcile" function does a lot of things, which means
it's hard to test properly, it's hard to follow the logic and
it's hard to handle all possible errors and events properly.

This splits reconcile into two parts, the first one that deals
preparing and validating the taskrun and all the resources it
depends on, the second one that actually reconciles the TaskRun
by creating the pod if required and updating the TaskRun from
the pod.

This adds events that were missing for error situation that
happens during preparation and validation.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [WIP] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
The TaskRun controller now emits events also when error occurs during the validation of the TaskRun or the resources it depends on.
```

### CODE REVIEW

File: pkg/reconciler/taskrun/taskrun.go [hunk 1]
```diff
@@ -182,38 +179,50 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
 	}
-	return multierror.Append(merr, c.updateStatusLabelsAndAnnotations(tr, original)).ErrorOrNil()
+	return multierror.Append(err, c.updateStatusLabelsAndAnnotations(tr, original)).ErrorOrNil()
 }
 
-func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error {
+// TODO(afrittoli) for now we return spec and resources. In future we should store
+// them in the TaskRun.Status so we don't need to re-run `prepare` at every
+// reconcile.
+func (c *Reconciler) prepare(ctx context.Context, tr *v1alpha1.TaskRun) (*v1alpha1.TaskSpec, *resources.ResolvedTaskResources, error) {
```
Comment (afrittoli) [inline pkg/reconciler/taskrun/taskrun.go:204]: Should these be *v1beta1 somehow?

File: pkg/reconciler/taskrun/taskrun.go [hunk 2]
```diff
@@ -275,8 +284,15 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 	}
 	cloudevent.InitializeCloudEvents(tr, prs)
 
+	return taskSpec, rtr, nil
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun,
```
Comment (afrittoli) [inline pkg/reconciler/taskrun/taskrun.go:290]: Should these be *v1beta1 somehow? After `prepare` they should be converted to beta?

Also I think I should move the emit event bit out of `reconcile` to `Reconcile`, so that all events are generated there.

Comment (vdemeester) [reply pkg/reconciler/taskrun/taskrun.go:290]: Nope. The are converted to `beta` and then to `alpha` to make sure they are in `beta` shape — but the internal (and stored) representation is still `v1alpha1`. But the `v1alpha1` struct you have after `prepare` are **sure** to be in the `v1beta1` shape, so you can/should/have-to use *only* `v1beta1` fields :angel: 

I know it's a bit confusing right now. This will change with #2413 and #2410 :angel:

Comment (afrittoli) [reply pkg/reconciler/taskrun/taskrun.go:300]: ack, thanks

File: pkg/reconciler/taskrun/taskrun.go [hunk 3]
```diff
@@ -182,38 +179,60 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 		return multierror.Append(err, c.updateStatusLabelsAndAnnotations(tr, original)).ErrorOrNil()
 	}
 
+	// prepare fetches all required resources, validates them together with the
```
Comment (sbwsg) [inline pkg/reconciler/taskrun/taskrun.go:182]: suggest this comment describing `prepare`'s behaviour be attached to the `func prepare(...)` on line 214.  But I do think it makes sense to leave a comment here about errors coming back from prepare being considered permanent.

File: pkg/reconciler/taskrun/taskrun.go (hunk 2 above)
Comment (sbwsg) [reply pkg/reconciler/taskrun/taskrun.go:300]: agree about event emitting being centralized into one area of the file 👍 

Also, could we add a comment about the conditions under which an error _should_ be returned by `func reconcile(...)`?  So often I see errors caught in the code but we `return nil` regardless.  I've never totally understood the logic of when to return an error from reconcile and when not to.

Comment (afrittoli) [reply pkg/reconciler/taskrun/taskrun.go:300]: I'm trying to make the logic about errors and events more consistent in the controller, so I will add some comments about this. 
One thing that I'd like to do eventually is also to start using `PermanentError`s, from knative/pkg. When `Reconcile` returns a `PermanentError` the key is not re-queued.

File: pkg/reconciler/taskrun/taskrun.go (hunk 3 above)
Comment (afrittoli) [reply pkg/reconciler/taskrun/taskrun.go:182]: Good point, will do

**Comment by afrittoli** [comment]
I cannot reproduce any of the integration test failures on my test cluster, even after multiple runs. I think that something in the logic that waits for events is not going well, not waiting long enough or perhaps starting to wait too soon.

**Comment by afrittoli** [comment]
> I cannot reproduce any of the integration test failures on my test cluster, even after multiple runs. I think that something in the logic that waits for events is not going well, not waiting long enough or perhaps starting to wait too soon.

I changed the cleanup logic so that namespaces are not cleaned up for failed tests. 
I can see all taskrun events, but the pipelinerun completion event is missing - I did not change that in my PR though - and I cannot reproduce the issue locally - with the only difference being that I only run one test at the time in my local environment.

**Comment by afrittoli** [comment]
I was finally able to reproduce the CI failures on my cluster. 
From local debugging, I see that after exactly 127 events, the API stops reporting events to clients. After 1h exactly, the initial 127 events expire, and magically the remaining events are returned to the client. I'm trying to find out if it is possible to change this behavior, else we might have to change the test to validate events only in unit tests.

**Comment by afrittoli** [comment]
> I was finally able to reproduce the CI failures on my cluster.
> From local debugging, I see that after exactly 127 events, the API stops reporting events to clients. After 1h exactly, the initial 127 events expire, and magically the remaining events are returned to the client. I'm trying to find out if it is possible to change this behavior, else we might have to change the test to validate events only in unit tests.

We'll have to use the `NewBroadcasterWithCorrelatorOptions` in the reconciler to tweak burst/qps options for the event broadcaster.

File: pkg/reconciler/taskrun/taskrun.go [hunk 4]
```diff
@@ -445,33 +467,31 @@ func (c *Reconciler) getTaskFunc(tr *v1alpha1.TaskRun) (resources.GetTask, v1alp
...
+			Type:    apis.ConditionSucceeded,
+			Status:  corev1.ConditionUnknown,
+			Reason:  podconvert.ReasonExceededResourceQuota,
+			Message: fmt.Sprintf("%s: %v", msg, err),
+		})
 	} else {
-		succeededStatus = corev1.ConditionFalse
-		reason = podconvert.ReasonCouldntGetTask
+		// The pod creation failed, not because of quota issues. The most likely
+		// reason is that something is wrong with the spec of the Task
```
Comment (dibyom) [inline pkg/reconciler/taskrun/taskrun.go:485]: > // reason is that something is wrong with the spec of the Task

shouldn't that be caught by validation beforehand?

File: test/init_test.go [hunk 5]
```diff
@@ -67,6 +67,16 @@ func setup(t *testing.T, fn ...func(*testing.T, *clients, string)) (*clients, st
 	return c, namespace
 }
 
+func setupStatic(t *testing.T) (*clients, string) {
+	t.Helper()
+	namespace := "faninfanout"
```
Comment (dibyom) [inline test/init_test.go:72]: I'm guessing this is for debugging only?

File: pkg/reconciler/taskrun/taskrun.go [hunk 6]
```diff
@@ -182,38 +179,60 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
+
+	// Emit events (only when ConditionSucceeded was changed)
+	after := tr.Status.GetCondition(apis.ConditionSucceeded)
+	reconciler.EmitEvent(c.Recorder, before, after, tr)
+
+	return multierror.Append(err, c.updateStatusLabelsAndAnnotations(tr, original)).ErrorOrNil()
 }
 
-func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error {
+// TODO(afrittoli) for now we return spec and resources. In future we should store
```
Comment (dibyom) [inline pkg/reconciler/taskrun/taskrun.go:211]: Would be nice for the TODO to reference a Github issue instead!

File: pkg/reconciler/taskrun/taskrun.go (hunk 4 above)
Comment (afrittoli) [reply pkg/reconciler/taskrun/taskrun.go:485]: Yeah, I did not change this logic here, and I had the same question.
My guess since we can have pod templates in taskrun, if something goes wrong in the pod creation if may be due to that... but tbh I don't think we can really know for sure.

File: test/init_test.go (hunk 5 above)
Comment (afrittoli) [reply test/init_test.go:72]: upz, forgot to remove this, thanks!

File: pkg/reconciler/taskrun/taskrun.go (hunk 6 above)
Comment (afrittoli) [reply pkg/reconciler/taskrun/taskrun.go:211]: I will add a link

File: pkg/reconciler/reconciler.go [hunk 7]
```diff
@@ -97,7 +97,13 @@ func NewBase(opt Options, controllerAgentName string, images pipeline.Images) *B
 	if recorder == nil {
 		// Create event broadcaster
 		logger.Debug("Creating event broadcaster")
-		eventBroadcaster := record.NewBroadcaster()
+
+		correlatorOptions := record.CorrelatorOptions{
+			// The default burst size is 25
+			BurstSize: 50,
```
Comment (dibyom) [inline pkg/reconciler/reconciler.go:103]: Probably fine for now but should be configurable right?

File: pkg/reconciler/reconciler.go [hunk 8]
```diff
@@ -97,7 +97,13 @@ func NewBase(opt Options, controllerAgentName string, images pipeline.Images) *B
...
+			QPS:       1,
```
Comment (dibyom) [inline pkg/reconciler/reconciler.go:104]: Is this the number of queries per second that we can emit? I couldn't figure out after reading the godoc for this

File: pkg/reconciler/reconciler.go (hunk 7 above)
Comment (afrittoli) [reply pkg/reconciler/reconciler.go:103]: Yes, I'll follow up on this

File: pkg/reconciler/reconciler.go (hunk 8 above)
Comment (afrittoli) [reply pkg/reconciler/reconciler.go:104]: Yes, the way I understood this is the QPS we allow after a burst
